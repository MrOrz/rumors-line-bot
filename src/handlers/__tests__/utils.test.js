import {
  createPostbackAction,
  createFeedbackWords,
  isNonsenseText,
  createReferenceWords,
  ellipsis,
} from '../utils';

describe('ellipsis()', () => {
  it('should not ellipsis when text is short', () => {
    expect(ellipsis('12345', 10)).toBe('12345');
  });

  it('should ellipsis when text is too long', () => {
    const limit = 5;
    const processed = ellipsis('1234567890', limit);
    expect(processed).toHaveLength(limit);
    expect(processed).toMatchSnapshot();
  });

  it('should properly cut emojis', () => {
    expect(ellipsis('🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈🏳️‍🌈', 5)).toMatchSnapshot();
  });

  it('should be able to customize ellipsis', () => {
    expect(ellipsis('1234567890', 5, '')).toBe('12345');
  });
});

describe('createPostbackAction()', () => {
  it('should return postback message body', () => {
    expect(
      createPostbackAction('閱讀此回應', 3, 1519019701265)
    ).toMatchSnapshot();
  });
});

describe('createFeedbackWords()', () => {
  it('should create empty feedback words', () => {
    expect(createFeedbackWords(0, 0)).toMatchSnapshot();
  });
  it('should create positive feedback words', () => {
    expect(createFeedbackWords(3, 0)).toMatchSnapshot();
  });
  it('should create negative feedback words', () => {
    expect(createFeedbackWords(0, 2)).toMatchSnapshot();
  });
  it('should create both feedback words', () => {
    expect(createFeedbackWords(1, 2)).toMatchSnapshot();
  });
});

describe('createReferenceWords()', () => {
  it('should create reference for rumors', () => {
    expect(
      createReferenceWords({
        reference: 'This is a reference',
        type: 'RUMOR',
      })
    ).toMatchSnapshot();
  });
  it('should create reference for opinions', () => {
    expect(
      createReferenceWords({
        reference: 'This is refering to different opinions',
        type: 'OPINIONATED',
      })
    ).toMatchSnapshot();
  });
});

// According to 20181017 meeting note, we chose to remove limitation temporarily.
xdescribe('isNonsenseText()', () => {
  it('should detect a info-less text [1]', () => {
    // https://cofacts.g0v.tw/article/AV84hj72yCdS-nWhuhxh
    let text = '這個人是不是在做援交的 請分析';
    expect(isNonsenseText(text)).toBe(true);
  });

  it('should detect a info-less text [2]', () => {
    // https://cofacts.g0v.tw/article/AV9MDX05yCdS-nWhuh73
    let text = '米飯 放涼 抗癌';
    expect(isNonsenseText(text)).toBe(true);
  });

  it('should detect a info-less text [3]', () => {
    // https://cofacts.g0v.tw/reply/njr6amQBbZnN2I-EfJfV
    let text = '嗨，請問蔡英文是同性戀，是真的嗎？';
    expect(isNonsenseText(text)).toBe(true);
  });

  it('should detect a normal text [1]', () => {
    let text = `如果你真的直接就去改資費方案的話，恐怕根本沒省到，
      甚至電話費還反而變貴了！https://today.line.me/TW/article/0BwyWj?utm_source=lineshare`;
    expect(isNonsenseText(text)).toBe(false);
  });

  it('should allow a normal text [2]', () => {
    let text = `http://goo.gl/3pZNro`;
    expect(isNonsenseText(text)).toBe(false);
  });

  /**
   *  Can't parse url without ^http.
   */
  xit('should detect a text with a single url', () => {
    let text = '不要點開 google.com';
    let newText = isNonsenseText(text);
    console.log({ text, newText });
  });
});
