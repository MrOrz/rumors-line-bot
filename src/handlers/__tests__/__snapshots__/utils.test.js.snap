// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SITE_URL use SITE_URL from env variables 1`] = `"https://cofacts.hacktabl.org/article/AWDZYXxAyCdS-nWhumlz"`;

exports[`Test SITE_URL use the default SITE_URL 1`] = `"https://cofacts.g0v.tw/article/AWDZYXxAyCdS-nWhumlz"`;

exports[`createArticleShareReply() should uri size less then 1000 1`] = `"line://msg/text/?%E6%88%91%E6%94%B6%E5%88%B0%E9%80%99%E5%89%87%E8%A8%8A%E6%81%AF%E7%9A%84%E6%83%B3%E6%B3%95%E6%98%AF%EF%BC%9A%0A%E6%88%91%E6%96%BC%E6%98%AF%E5%90%91%E4%BB%96%E8%AA%AA%3A%E3%80%8C%E4%BD%A0%E4%B9%9F%E7%9F%A5%E9%81%93%E9%82%A3%E6%B1%9F%E6%B0%B4%E5%92%8C%E6%9C%88%E4%BA%AE%E5%97%8E%EF%BC%9F%E6%B1%9F%E6%B0%B4%E9%9B%96%E7%84%B6%E4%B8%8D%E5%81%9C%E7%9A%84%E6%B5%81%EF%BC%8C%E4%BD%86%E6%98%AF%E5%AE%83%E7%9A%84%E6%9C%AC%E9%AB%94%E4%B8%A6%E4%B8%8D%E6%9B%BE%E6%B5%81%E5%8E%BB%E5%95%8A%3B%E6%9C%88%E4%BA%AE%E9%9B%96%E7%84%B6%E6%9C%89%E5%9C%93%E6%9C%89%E7%BC%BA%EF%BC%8C%E5%8F%AF%E6%98%AF%E6%9C%88%E7%9A%84%E6%9C%AC%E8%BA%AB%E4%B8%A6%E6%B2%92%E6%9C%89%E5%A2%9E%E6%B8%9B%E5%95%8A%E3%80%82%E8%8B%A5%E5%BE%9E%E8%AE%8A%E7%9A%84%E2%8B%AF%E2%8B%AF%0A%0A%E8%AB%8B%E5%B9%AB%E6%88%91%E7%9C%8B%E7%9C%8B%E9%80%99%E6%98%AF%E7%9C%9F%E7%9A%84%E9%82%84%E6%98%AF%E5%81%87%E7%9A%84%EF%BC%9Ahttps%3A%2F%2Fcofacts.hacktabl.org%2Farticle%2FAWDZYXxAyCdS-nWhumlz"`;

exports[`createArticleShareReply() should uri size less then 1000 2`] = `"https://www.facebook.com/dialog/share?openExternalBrowser=1&app_id=719656818195367&display=popup&quote=%E6%88%91%E6%96%BC%E6%98%AF%E5%90%91%E4%BB%96%E8%AA%AA%3A%E3%80%8C%E4%BD%A0%E4%B9%9F%E7%9F%A5%E9%81%93%E9%82%A3%E6%B1%9F%E6%B0%B4%E5%92%8C%E6%9C%88%E4%BA%AE%E5%97%8E%EF%BC%9F%E6%B1%9F%E6%B0%B4%E9%9B%96%E7%84%B6%E4%B8%8D%E5%81%9C%E7%9A%84%E6%B5%81%EF%BC%8C%E4%BD%86%E6%98%AF%E5%AE%83%E7%9A%84%E6%9C%AC%E9%AB%94%E4%B8%A6%E4%B8%8D%E6%9B%BE%E6%B5%81%E5%8E%BB%E5%95%8A%3B%E6%9C%88%E4%BA%AE%E9%9B%96%E7%84%B6%E6%9C%89%E5%9C%93%E6%9C%89%E7%BC%BA%EF%BC%8C%E5%8F%AF%E6%98%AF%E6%9C%88%E7%9A%84%E6%9C%AC%E8%BA%AB%E4%B8%A6%E6%B2%92%E6%9C%89%E5%A2%9E%E6%B8%9B%E5%95%8A%E3%80%82%E8%8B%A5%E5%BE%9E%E8%AE%8A%E7%9A%84%E8%A7%92%E5%BA%A6%E4%BE%86%E7%9C%8B%EF%BC%8C%E9%82%A3%E9%BA%BC%E5%A4%A9%E5%9C%B0%E9%96%93%E2%8B%AF%E2%8B%AF&hashtag=%23Cofacts%E6%B1%82%E8%A7%A3%E6%83%91&href=https%3A%2F%2Fcofacts.hacktabl.org%2Farticle%2FAWDZYXxAyCdS-nWhumlz"`;

exports[`createFeedbackWords() should create both feedback words 1`] = `
"[有 1 人覺得此回應有幫助
有 2 人覺得此回應沒幫助]"
`;

exports[`createFeedbackWords() should create empty feedback words 1`] = `"[還沒有人針對此回應評價]"`;

exports[`createFeedbackWords() should create negative feedback words 1`] = `"[有 2 人覺得此回應沒幫助]"`;

exports[`createFeedbackWords() should create positive feedback words 1`] = `"[有 3 人覺得此回應有幫助]"`;

exports[`createFlexMessageText should create a text for flex message 1`] = `"計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入，下車時使用老人悠遊卡，跳錶車資105元，優惠32元，只扣73元，哈哈，這是屬於我們的福利，與大家分享，可以善加利用！=7"`;

exports[`createFlexMessageText should handle the situation without input 1`] = `""`;

exports[`createPostbackAction() should return postback message body 1`] = `
Object {
  "data": "{\\"input\\":3,\\"issuedAt\\":1519019701265}",
  "label": "閱讀此回應",
  "type": "postback",
}
`;

exports[`createReferenceWords() should create reference for opinions 1`] = `"不同觀點請見：This is refering to different opinions"`;

exports[`createReferenceWords() should create reference for rumors 1`] = `"出處：This is a reference"`;

exports[`createTypeWords should return the type words for NOT_ARTICLE 1`] = `"不在查證範圍"`;

exports[`createTypeWords should return the type words for NOT_RUMOR 1`] = `"含有真實訊息"`;

exports[`createTypeWords should return the type words for OPINIONATED 1`] = `"含有個人意見"`;

exports[`createTypeWords should return the type words for RUMOR 1`] = `"含有不實訊息"`;

exports[`createTypeWords should return the type words for other types 1`] = `"回應的狀態未定義！"`;

exports[`ellipsis() should ellipsis when text is too long 1`] = `"123⋯⋯"`;

exports[`ellipsis() should properly cut emojis 1`] = `"🏳️‍🌈🏳️‍🌈🏳️‍🌈⋯⋯"`;
